plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/VectorFunctions.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.postgresql;
+
+import io.trino.spi.TrinoException;
+import io.trino.spi.block.Block;
+import io.trino.spi.function.Description;
+import io.trino.spi.function.ScalarFunction;
+import io.trino.spi.function.SqlType;
+
+import static io.trino.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static io.trino.spi.type.RealType.REAL;
+import static io.trino.spi.type.StandardTypes.DOUBLE;
+
+public final class VectorFunctions
+{
+    private VectorFunctions() {}
+
+    @Description("Calculates the euclidean distance between two vectors. This is same as L2 distance <-> in pgvector.")
+    @ScalarFunction
+    @SqlType(DOUBLE)
+    public static double euclideanDistance(@SqlType("array(real)") Block first, @SqlType("array(real)") Block second)
+    {
+        checkArgument(first.getPositionCount() == second.getPositionCount(), "The arguments must have the same length");
+
+        double sum = 0.0;
+        for (int i = 0; i < first.getPositionCount(); i++) {
+            double firstValue = REAL.getFloat(first, i);
+            double secondValue = REAL.getFloat(second, i);
+            sum = sum + Math.pow((firstValue - secondValue), 2.0);
+        }
+        return Math.sqrt(sum);
+    }
+
+    @Description("Calculates the inner product between two vectors. This is same as cosine distance <#> in pgvector.")
+    @ScalarFunction
+    @SqlType(DOUBLE)
+    public static double innerProduct(@SqlType("array(real)") Block first, @SqlType("array(real)") Block second)
+    {
+        checkArgument(first.getPositionCount() == second.getPositionCount(), "The arguments must have the same length");
+
+        double[] firstVector = new double[first.getPositionCount()];
+        double[] secondVector = new double[second.getPositionCount()];
```

Comment:
We could avoid allocating these arrays by doing the computation inline instead of calling out to `dotProduct`. Can we add some JMH benchmarks to assess which way is better ?

Reply:
Added `BenchmarkArrayInnerProduct`. The new average time is the half of the old implementation.